### PR TITLE
DE-6746: Add a GH workflow to publish a release

### PR DESCRIPTION
**Pull request by fingerartur**
Here I'm adding a GitHub workflow to publish a new version to NPM and deploy API docs for that version to https://players.castlabs.com/react-dom/x.x.x/docs/. It gets triggered when we create a new Release in GitHub (when creating the release, select "Create new tag" and specify the target version).

Beta versions are also supported and they will be published accordingly to a NPM beta channel (non-main channel).

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        run: aws s3 cp --recursive "./dist/storybook" "${{ secrets.AWS_WEB_BUCKET }}/${{ env.WEB_PATH }}"
+        # Note: the default cache policy is 1 day, I think that's OK for now.
+        # In the future if traffic increases we can increase the expiration
+        # as much as we want.
+
+    ## Publish the NPM package
+
+      - name: Write NPM RC File
+        run: |
+          echo '@castlabs:registry=https://registry.npmjs.org' > .npmrc
```
Comment (thasso) [inline .github/workflows/publish.yml:59]: This one gives me unease :) Can we 1) Use the secret directly here and 2) add a finally step that always executes and removes the .npmrc file?

File: story/stories/examples/DefaultSkinSource.tsx [hunk 2]
```diff
@@ -14,7 +15,7 @@ import '@castlabs/prestoplay-react-components/dist/themes/pp-ui-base-theme-embed
 
 const baseConfig = {
   // Replace this by your license key
-  license: '7c8165eb7231508ce4eae02181d53b1c',
+  license: '7c816...81d53b1c',
```
Comment (thasso) [inline story/stories/examples/DefaultSkinSource.tsx:18]: Maybe we add a comment here that leaving this out will use the embedded dev license or even make it empty and just have the comment so it "just works" for local development?

File: story/stories/examples/YoutubeSkinSource.ts [hunk 3]
```diff
@@ -32,16 +24,14 @@ import 'youtube-style.css'
...
-const asset =   {
-  title: 'Maldives',
-  subtitle: 'State of the Mahal Dibiyat',
-  config: {
-    source: 'https://content.players.castlabs.com/demos/clear-segmented/manifest.mpd',
-  },
+
+const baseConfig = {
+  // Replace this by your license key
+  license: '7c8165...81d53b1c',
```
Comment (thasso) [inline story/stories/examples/YoutubeSkinSource.ts:30]: s.o.

File: .github/actions/publish/main.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+const regexBetaVersion = /^\d+\.\d+\.\d+-(\w+)\.\d+$/;
+const regexVersion = /^\d+\.\d+\.\d+$/;
+
+const betaMatch = version.match(regexBetaVersion);
+
+if (betaMatch) {
+  const betaLabel = betaMatch[1];
+  core.info(`Publishing a Beta version ${version}.`)
+  execSync(`npm publish ${dryRun ? '--dry-run' : ''} --tag ${betaLabel}`, { stdio: 'inherit' })
+  process.exit(0);
```
[the remark this answers is not in the document]
Comment (fingerartur) [reply .github/actions/publish/main.js:21]: Ok I'll get rid of those process.exit(0);

File: .github/workflows/publish.yml (hunk 1 above)
Comment (fingerartur) [reply .github/workflows/publish.yml:59]: Yeah, I'll pass it directly to the shell that needs it.

File: story/stories/examples/YoutubeSkinSource.ts (hunk 3 above)
Comment (fingerartur) [reply story/stories/examples/YoutubeSkinSource.ts:30]: Fair warning: My German language skills suck incredibly 😄

File: .github/workflows/publish.yml (hunk 1 above)
Comment (fingerartur) [reply .github/workflows/publish.yml]: Actually ... I've been bamboozled. I think we both may have misunderstood what this step does - it does **_not_** write out the NPM_TOKEN to that file, it creates a templated .npmrc file, and NPM then fills that template with the current value of `$NPM_TOKEN` at runtime, meaning the token is in RAM all that time, it does not get written to a file. ... So I think we are happy with this step security-wise.

So I will keep the logic of this step, but I will move it to the `.github/actions/publish/main.js` action, which does the publishing to NPM.

**Comment by fingerartur** [comment]
@thasso I've also added provenance verification and switched to a granular NPM token.

Provenance works, I've tested it with a beta NPM release, just an interesting note on that topic: it needs Node.js version 20 (actually version 18 would be enough, but GitHub actions somehow don't support it, at least not explicitly, so version 20 it is 😄).

File: .github/actions/publish/main.js [hunk 5]
```diff
@@ -1,33 +1,72 @@
...
+core.info(`Publishing ${type === 'beta' ? 'Beta version' : 'version'} ${version}.`)
+
+const npmRcContent = `
+@castlabs:registry=https://registry.npmjs.org
+//registry.npmjs.org/:_authToken=\${NPM_TOKEN}
+`
+
+try {
+  fs.writeFileSync(utils.file('.npmrc'), npmRcContent)
+  execSync(`export NPM_TOKEN=${npmToken}; npm publish ${args.join(' ')}`, { stdio: 'inherit' })
```
Comment (thasso) [inline .github/actions/publish/main.js:68]: I think you do not even need to `export` the token here. I think you can either _pass_ this for that one shell commend, i.e. `NPM_TOKEN="" npm publish...` or (and I tend to like that more) pass it as an `env` option in the `options` of `execSync`.

**Review by thasso — CHANGES_REQUESTED** [review]
nice, just one more suggestion on how we expose the token that you might want to give a shot.

File: .github/actions/publish/main.js (hunk 5 above)
Comment (fingerartur) [reply .github/actions/publish/main.js]: Nice, I will use `env`.

**Review by thasso — APPROVED** [review]

**Pull request merged**